### PR TITLE
Support MFA with backup codes

### PR DESCRIPTION
**Pull request by natalian98**
This PR adds support for lookup secrets (backup codes) as a fallback method for TOTP multi-factor auth.
--- ---
Setting up the codes (`/ui/setup_backup_codes`):
image
image
Print/download:
image
image

image

Sign in with a backup code:
image
image
Error messages:
image
If user logged in using a backup code and has 3 or less of them left, the following screen will be displayed, offering to generate a new set:
image

### CODE REVIEW

File: pkg/kratos/handlers.go [hunk 1]
```diff
@@ -206,10 +207,56 @@ func (a *API) handleUpdateFlow(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		if a.is40xError(err) {
+			a.logger.Debugf("check session failed: %v", err)
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	aal2AuthenticationMethod := session.AuthenticationMethods[1].Method
```
Comment (BarcoMasile) [inline pkg/kratos/handlers.go:245]: question: is index 1 here because we're at the second step of the mfa-enforced login?

File: pkg/kratos/service.go [hunk 2]
```diff
@@ -657,6 +696,47 @@ func (s *Service) HasTOTPAvailable(ctx context.Context, id string) (bool, error)
...
+	identity, _, err := s.kratosAdmin.IdentityApi().
+		GetIdentity(ctx, id).
+		IncludeCredential([]string{"lookup_secret"}).
+		Execute()
+
+	if err != nil {
+		return false, err
+	}
+
+	lookupCredentials := identity.GetCredentials()["lookup_secret"].Config["recovery_codes"]
```
Comment (BarcoMasile) [inline pkg/kratos/service.go:710]: possibile issue: do we know for certain that `"recovery_codes"` key will always be available?
If that's not the case, I'll put my suggestion here:

```go
lookUpSecret, ok := identity.GetCredentials()["lookup_secret"]
if !ok {
    // handle case key is not present
}

lookupCredentials, ok := lookUpSecret.Config["recovery_codes"]
if !ok {
    // handle case key is not present
}
```

File: pkg/kratos/service.go [hunk 3]
```diff
@@ -657,6 +696,47 @@ func (s *Service) HasTOTPAvailable(ctx context.Context, id string) (bool, error)
...
+	if err != nil {
+		s.logger.Errorf("Marshalling to json failed: %s", err)
+		return false, err
+	}
+
+	lookupSecrets := new(LookupSecrets)
+	if err := json.Unmarshal(jsonbody, &lookupSecrets); err != nil {
+		s.logger.Errorf("Unmarshalling failed: %s", err)
+		return false, err
+	}
```
Comment (BarcoMasile) [inline pkg/kratos/service.go:722]: suggestion(non-blocking): how about switching to the `json.NewDecoder` approach, so it's leaner ?

File: pkg/kratos/service.go [hunk 4]
```diff
@@ -657,6 +696,47 @@ func (s *Service) HasTOTPAvailable(ctx context.Context, id string) (bool, error)
...
+	unusedCodes := 0
+	for _, code := range *lookupSecrets {
+		if code.UsedAt.IsZero() {
+			unusedCodes += 1
+		}
+	}
+
+	if unusedCodes > 3 {
+		return false, nil
+	}
```
Comment (BarcoMasile) [inline pkg/kratos/service.go:733]: suggestion: this is great already, but I suggest you create at least a function to isolate this specific logic of determining if codes available are enough or not. I'm referring to lines 731-733

File: ui/components/Flow.tsx [hunk 5]
```diff
@@ -21,7 +21,14 @@ export type Values = Partial<
...
 
-export type Methods = "oidc" | "password" | "code" | "totp" | "webauthn";
+export type Methods =
+  | "oidc"
+  | "password"
+  | "code"
+  | "totp"
+  | "webauthn"
+  | "lookup"
+  | "lookup_secret";
```
Comment (BarcoMasile) [inline ui/components/Flow.tsx:31]: question: is "lookup" an alias for "lookup_secret"?

File: ui/components/NodeInputSubmit.tsx [hunk 6]
```diff
@@ -33,39 +33,55 @@ export const NodeInputSubmit: FC<NodeInputProps> = ({
...
+        ) : (
+          label
+        )}
+      </Button>
+      {showBackupLink && (
+        <Button
+          appearance="link"
+          type="button"
+          onClick={() =>
+            (window.location.href = window.location.href + "&use_backup_code=1")
```
Comment (BarcoMasile) [inline ui/components/NodeInputSubmit.tsx:81]: question: I would expect a `true` value instead of `1`, just curious to know why we have an int here

File: ui/components/NodeText.tsx [hunk 7]
```diff
@@ -18,23 +18,50 @@ const Content: FC<Props> = ({ attributes }) => {
...
+                  const element = document.createElement("a");
+                  const file = new Blob([text], {
+                    type: "text/plain",
+                  });
+                  element.href = URL.createObjectURL(file);
+                  element.download = "backup-codes.txt";
+                  document.body.appendChild(element);
+                  element.click();
+                  document.body.removeChild(element);
+                }}
```
Comment (BarcoMasile) [inline ui/components/NodeText.tsx]: suggestion: I have to ask why inlining this whole function here? Why not relying on `useCallback` ?
Same applies at line 53

File: ui/components/NodeText.tsx [hunk 8]
```diff
@@ -29,21 +31,27 @@ const Content: FC<Props> = ({ attributes }) => {
...
-                  const element = document.createElement("a");
-                  const file = new Blob([text], {
-                    type: "text/plain",
-                  });
-                  element.href = URL.createObjectURL(file);
-                  element.download = "backup-codes.txt";
-                  document.body.appendChild(element);
-                  element.click();
-                  document.body.removeChild(element);
+                onClick={async () => {
```
Comment (BarcoMasile) [inline ui/components/NodeText.tsx:44]: same suggestoin here to use `useCallback`

**Review by BarcoMasile — COMMENTED** [review]
Great work overall, some suggestions

File: ui/components/Flow.tsx (hunk 5 above)
Comment (natalian98) [reply ui/components/Flow.tsx]: Only `lookup_secret` is valid, I removed the `lookup`.

File: pkg/kratos/handlers.go [hunk 9]
```diff
@@ -279,6 +326,27 @@ func (a *API) mfaSettingsRedirect(w http.ResponseWriter) {
 	)
 }
 
+func (a *API) lookupSecretsSettingsRedirect(w http.ResponseWriter, flowId string) {
+	redirectUrl, err := url.Parse(path.Join("/", a.contextPath, "/ui/backup_codes_regenerate?flow="+flowId))
```
Comment (nsklikas) [inline pkg/kratos/handlers.go:330]: nit-pick: use the `UI` constant found in the `ui` package instead of hardcoding the `ui` path?

File: pkg/kratos/handlers.go [hunk 10]
```diff
@@ -279,6 +326,27 @@ func (a *API) mfaSettingsRedirect(w http.ResponseWriter) {
...
+	if err != nil {
+		err = fmt.Errorf("unable to build backup codes redirect path, possible misconfiguration, err: %v", err)
+		a.logger.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	redirect := redirectUrl.String()
+	errorId := "regenerate_backup_codes"
```
Comment (nsklikas) [inline pkg/kratos/handlers.go:339]: suggestion: Move the `"regenerate_backup_codes"` to a global constant.

File: pkg/kratos/handlers.go [hunk 11]
```diff
@@ -206,10 +207,56 @@ func (a *API) handleUpdateFlow(w http.ResponseWriter, r *http.Request) {
...
+
+		a.logger.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if shouldRegenerateBackupCodes {
+		a.lookupSecretsSettingsRedirect(w, flowId)
+		return
+	}
```
Comment (nsklikas) [inline pkg/kratos/handlers.go:222]: question: why do we set the cookies before we apply this logic? If we get an error, we will return a 500 but still initialize a session, which feels a little counter intuitive for me.

File: pkg/kratos/handlers.go [hunk 12]
```diff
@@ -206,10 +207,56 @@ func (a *API) handleUpdateFlow(w http.ResponseWriter, r *http.Request) {
...
 	}
 
+	shouldRegenerateBackupCodes, err := a.shouldRegenerateBackupCodes(r.Context(), cookies)
+	if err != nil {
+		err = fmt.Errorf("error when checking backup codes: %v", err)
+
+		a.logger.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```
Comment (nsklikas) [inline pkg/kratos/handlers.go:217]: Not sure if I understand correctly, so please correct me if I'm wrong:

This endpoint may be called multiple times during a user's login flow, once for the username/password, once for OTP, once for webauthn. This means that this logic will run multiple times and if I don't want to print new codes (which is not unlikely) I will have to decline this every time. 

I have not run it to see if this is the actual behavior, so I may be missing something.

**Review by nsklikas — COMMENTED** [review]
code looks fine (didn't look to much into the frontend), but I have a some questions and some nit-picks

File: pkg/kratos/handlers.go (hunk 12 above)
Comment (natalian98) [reply pkg/kratos/handlers.go:217]: The user will only be offered to regenerate the codes if the method used to perform MFA was lookup secret and there are 3 or less codes left.
The screen is not presented multiple times (there are checks in the `shouldRegenerateBackupCodes` handler to prevent this) so the user can only decline it once.

File: pkg/kratos/service.go (hunk 3 above)
Comment (natalian98) [reply pkg/kratos/service.go:733]: Agreed in private that it's not worth to change it to maintain readability

File: pkg/kratos/service.go (hunk 2 above)
Comment (natalian98) [reply pkg/kratos/service.go]: Technically yes, because this check is only executed if the user logged in with `lookup_secret` method, meaning the identity has recovery codes. But I added the additional checks in case anything goes wrong.

File: pkg/kratos/handlers.go (hunk 12 above)
Comment (nsklikas) [reply pkg/kratos/handlers.go:212]: Ah, I missed that thanks. What if we have more than 2fa though? Is this possible to happen (passkey as 3rd fa or something like that)

Comment (natalian98) [reply pkg/kratos/handlers.go:212]: Passkey can only be used as the first factor (this is restricted by kratos settings). No matter how many factors are there, this screen will only be shown once when a backup code is used.

File: ui/components/NodeInputSubmit.tsx (hunk 6 above)
Comment (edlerd) [reply ui/components/NodeInputSubmit.tsx:81]: changing it to true.

File: ui/components/NodeText.tsx (hunk 8 above)
Comment (edlerd) [reply ui/components/NodeText.tsx:44]: Extracted and wrapped with use callback.

**Review by BarcoMasile — APPROVED** [review]
I'll approve so we can merge, there are still a few inline labmdas to swtich to the `useCallback` approach, but that'll be at a different time
Thank you

**Review by nsklikas — DISMISSED** [review]

File: pkg/kratos/service.go [hunk 13]
```diff
@@ -657,6 +697,57 @@ func (s *Service) HasTOTPAvailable(ctx context.Context, id string) (bool, error)
 	return ok, nil
 }
 
+func (s *Service) HasNotEnoughLookupSecretsLeft(ctx context.Context, id string) (bool, error) {
```
Comment (shipperizer) [inline pkg/kratos/service.go:700]: this method has no tests

File: pkg/kratos/handlers.go [hunk 14]
```diff
@@ -199,17 +203,69 @@ func (a *API) handleUpdateFlow(w http.ResponseWriter, r *http.Request) {
...
+
 	setCookies(w, cookies)
 
 	if shouldEnforceMfa {
```
Comment (shipperizer) [inline pkg/kratos/handlers.go:217]: not part of this pr, but there are not tests covering this if clause

mind adding an issue in gh?

Comment (natalian98) [reply pkg/kratos/handlers.go:217]: Sure
https://github.com/canonical/identity-platform-login-ui/issues/271

File: pkg/kratos/handlers.go [hunk 15]
```diff
@@ -279,6 +335,27 @@ func (a *API) mfaSettingsRedirect(w http.ResponseWriter) {
 	)
 }
 
+func (a *API) lookupSecretsSettingsRedirect(w http.ResponseWriter, flowId string) {
+	redirectUrl, err := url.Parse(path.Join("/", a.contextPath, ui.UI, "/backup_codes_regenerate?flow="+flowId))
```
[the remark this answers is not in the document]
Comment (natalian98) [reply pkg/kratos/handlers.go:339]: Yes, because it's a subpath

File: pkg/kratos/service.go (hunk 13 above)
Comment (natalian98) [reply pkg/kratos/service.go:700]: Tests added

**Review by BarcoMasile — APPROVED** [review]

File: pkg/kratos/handlers.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (shipperizer) [reply pkg/kratos/handlers.go:339]: subpath as front-end double slash path?

**Review by shipperizer — APPROVED** [review]
